### PR TITLE
feat: comment endpoints

### PR DESCRIPTION
**Pull request by ZainabImadulla**
# Description

Implemented and tested comment endpoints and adding documentation to the openapi.yaml file.

Created Endpoints:

POST /posts/:id/comments: Add a comment to a post
DELETE /comments/:id: Delete a comment
GET /posts/:id/comments: Get all comments under the post (paginated)

Also added in an additional test method in TestBuilder that allows for testing specific fields for a response body that is an array. 
## Checklist:

- [X] I have performed a self-review of my code.
- [X] I have added necessary tests (unit, integration, etc.) to cover the new functionality or fix.
- [X] All new and existing tests pass locally.
- [X] I have updated the documentation to reflect changes (e.g., README, code comments).
- [X] I have removed any commented-out code, debug statements, or unnecessary console logs.

### CODE REVIEW

File: backend/src/entities/comments/transaction.ts [hunk 1]
```diff
@@ -59,4 +67,87 @@ export class CommentTransactionImpl implements CommentTransaction {
       }
     });
   }
+
+  async createComment(comment: CreateCommentPayload): Promise<Comment | null> {
+    const createdComment = await this.db.transaction(async (tx) => {
```
Comment (in-mai-space) [inline backend/src/entities/comments/transaction.ts:72]: this is only one query so we don't need to wrap it in a transaction, we can use transaction only when there is multiple queries and we want to ensure that all fail if one operation fail

File: backend/src/tests/comments/getcomments.test.ts [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+  const goodRequestBody = {
+    content: "i like this photo",
+  };
+
+  beforeAll(async () => {
+    app = await startTestApp();
+  });
+
+  it("should return 200 if comments are successfully fetched", async () => {
```
Comment (in-mai-space) [inline backend/src/tests/comments/getcomments.test.ts:24]: can we also test with pagination as well to ensure it returns correctly with limit and page change? you can also take a look at GET members pagination test!

**Review by stoneliuCS — CHANGES_REQUESTED** [review]
Nice work! Review some of my comments and make the necessary changes!

File: backend/src/entities/comments/service.ts [hunk 3]
```diff
@@ -22,4 +30,25 @@ export class CommentServiceImpl implements CommentService {
...
+  async createComment(payload: CreateCommentPayload): Promise<Comment | null> {
+    const createCommentImpl = async () => {
+      return await this.commentTransaction.createComment(payload);
+    };
+    return await handleServiceError(createCommentImpl)();
+  }
+
+  async getComments(payload: CommentPagination): Promise<Comment[]> {
+    const getCommentsImpl = async () => {
+      return await this.commentTransaction.getComments(payload);
```
Comment (in-mai-space) [inline backend/src/entities/comments/service.ts:43]: if a comment is a voice memo, we will need to use the objectKey (uuid) to retrieve the signed url that you and Josh made! in this case, there's a media service that is responsible for doing this, so we can inject the mediaService into the commentService as well. you can take a look at getMembers in MemberService!

**Review by in-mai-space — CHANGES_REQUESTED** [review]
@ZainabImadulla love the implementation so far and great tests for edge cases! i just left some comments, after you resolve them we can merge and deploy! Let me know if you have any questions 🐐🔥✨

File: backend/src/entities/comments/service.ts (hunk 3 above)
Comment (ZainabImadulla) [reply backend/src/entities/comments/service.ts:43]: So just to clarify, when getting the comments we want to return the url of the voice memo rather than the objectKey?

Comment (in-mai-space) [reply backend/src/entities/comments/service.ts:43]: yup! because we want users in the frontend to be able to actually play the audio, so returning the objectKey won't help much. 

I know that we can also take the objectKey and do another separate GET request for the url, but it will take more time since we have to call another endpoint and that can affect comment loading speed, so retrieving the url when retrieving the comments will be better approach!

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -0,0 +1,87 @@
+name: Deployment
+
+env:
+  REGISTRY: docker.io
+  IMAGE_NAME: liusto258/dearly 
+  DOCKER_IMAGE_URL: liusto258/dearly:main
+  SERVICE: dearly
```
[the remark this answers is not in the document]
Comment (in-mai-space) [reply .github/workflows/build.yml:7]: @ZainabImadulla yeh i believe this file is now called .github/workflows/deployment.yml, probably get lost when resolving conflict ^

File: backend/src/entities/comments/transaction.ts [hunk 5]
```diff
@@ -59,4 +72,65 @@ export class CommentTransactionImpl implements CommentTransaction {
       }
     });
   }
+
+  async createComment(comment: CreateCommentPayload): Promise<Comment> {
+    const [newComment] = await this.db.insert(commentsTable).values(comment).returning();
+    if (!newComment) throw new InternalServerError();
+    return newComment;
+  }
```
Comment (in-mai-space) [inline backend/src/entities/comments/transaction.ts:80]: Can we also check whether user is in the group where the post is made? If user is not a member in the group, we should throw 403 Forbidden since they should not be able to comment on the post.

For InternalServerError, can we also add the argument "Failed to create comment" just to make the error more informative".

File: backend/src/tests/comments/getcomments.test.ts [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+      ],
+    ],
+    ["1", "3", []],
+    ["1", "4", []],
+    [
+      "2",
+      "1",
+      [
+        {
+          content: "amazing photos!",
```
Comment (in-mai-space) [inline backend/src/tests/comments/getcomments.test.ts:54]: Can we make these into constants in test suites like BOB_COMMENT so it's not repeated and look neater?

File: backend/src/tests/comments/post.test.ts [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+
+describe("POST /posts/:id/comments", () => {
+  let app: Hono;
+  const testBuilder = new TestBuilder();
+  const ANA_JWT = generateJWTFromID(USER_ANA_ID);
+  const goodRequestBody = {
+    content: "i like this photo",
+  };
+  const goodRequestBody2 = {
+    voiceMemo: randomUUIDv7(),
```
Comment (in-mai-space) [inline backend/src/tests/comments/post.test.ts:17]: Very small nit, we do have a function called `generateUUID`, can we change randomUUIDv7 to `generateUUID` just so it is consistent with all other tests?

File: backend/src/entities/comments/service.ts [hunk 8]
```diff
@@ -22,4 +30,25 @@ export class CommentServiceImpl implements CommentService {
     };
     return await handleServiceError(toggleLikeCommentImpl)();
   }
+
+  async createComment(payload: CreateCommentPayload): Promise<Comment | null> {
```
Comment (in-mai-space) [inline backend/src/entities/comments/service.ts:34]: Since now it's guaranteed that we will always return a comment, we can remove the null from the Promise!

File: backend/src/entities/schema.ts [hunk 9]
```diff
@@ -120,7 +120,7 @@ export const commentsTable = pgTable("comments", {
     .notNull()
     .references(() => postsTable.id, { onDelete: "cascade" }),
   content: varchar({ length: TEXT_MAX_LIMIT }),
-  voiceMemo: varchar(),
+  voiceMemo: uuid(),
```
Comment (in-mai-space) [inline backend/src/entities/schema.ts:123]: can we run `task db:generate` so it generates a new SQL migration file? We need a new migration file everytime we change the database schema.

**Review by in-mai-space — APPROVED** [review]
@ZainabImadulla  ✨LGTM 🐐 TO PRODUCTION WE GO 🚀
